wasm2c/examples/segue/main.c

```diff
@@ -0,0 +1,265 @@
+                                         u32 b,
+                                         u64 c,
+                                         u32 d,
+                                         u32 e) {
+  printf("fd_seek not implemented\n");
+  abort();
+}
+u32 w2c_wasi__snapshot__preview1_fd_read(struct w2c_wasi__snapshot__preview1* a,
+                                         u32 b,
+                                         u32 c,
+                                         u32 d,
+                                         u32 e) {
+  printf("fd_read not implemented\n");
+  abort();
+}
+u32 w2c_wasi__snapshot__preview1_fd_close(
+    struct w2c_wasi__snapshot__preview1* a,
+    u32 b) {
+  printf("fd_close not implemented\n");
+  abort();
+}
+u32 w2c_wasi__snapshot__preview1_fd_fdstat_set_flags(
+    struct w2c_wasi__snapshot__preview1* a,
+    u32 b,
+    u32 c) {
+  printf("fd_fdstat_set_flags not implemented\n");
+  abort();
+}
+u32 w2c_wasi__snapshot__preview1_fd_prestat_dir_name(
+    struct w2c_wasi__snapshot__preview1* a,
+    u32 b,
+    u32 c,
+    u32 d) {
+  printf("fd_prestat_dir_name not implemented\n");
+  abort();
+}
+u32 w2c_wasi__snapshot__preview1_path_open(
+    struct w2c_wasi__snapshot__preview1* a,
+    u32 b,
+    u32 c,
+    u32 d,
+    u32 e,
+    u32 f,
+    u64 g,
+    u64 h,
+    u32 i,
+    u32 end) {
+  printf("path_open not implemented\n");
+  abort();
+}
+void w2c_wasi__snapshot__preview1_proc_exit(
+    struct w2c_wasi__snapshot__preview1* a,
+    u32 b) {
+  printf("proc_exit not implemented\n");
+  abort();
+}
+
+int main(int argc, char const* argv[]) {
+  w2c_dhrystone dhrystone;
+  struct w2c_wasi__snapshot__preview1 wasi;
+  uvwasi_t local_uvwasi_state;
+  uvwasi_options_t init_options;
+
+  // pass in standard descriptors
+  init_options.in = 0;
+  init_options.out = 1;
+  init_options.err = 2;
+  init_options.fd_table_size = 10;
+
+  // pass in args and environement
+  extern const char** environ;
+  init_options.argc = argc;
+  init_options.argv = argv;
+  init_options.envp = (const char**)environ;
+
+  // no sandboxing enforced, binary has access to everything user does
+  init_options.preopenc = 2;
+  init_options.preopens = calloc(2, sizeof(uvwasi_preopen_t));
+
+  init_options.preopens[0].mapped_path = "/";
+  init_options.preopens[0].real_path = "/";
+  init_options.preopens[1].mapped_path = "./";
+  init_options.preopens[1].real_path = ".";
+
+  init_options.allocator = NULL;
+
+  wasm_rt_init();
+  uvwasi_errno_t ret = uvwasi_init(&local_uvwasi_state, &init_options);
+
+  if (ret != UVWASI_ESUCCESS) {
+    printf("uvwasi_init failed with error %d\n", ret);
+    exit(1);
+  }
+
+  wasi.w2c_memory = &dhrystone.w2c_memory;
+  wasi.uvwasi = &local_uvwasi_state,
+
+  wasm2c_dhrystone_instantiate(&dhrystone, &wasi);
+
+  w2c_dhrystone_0x5Fstart(&dhrystone);
+
+  wasm2c_dhrystone_free(&dhrystone);
+
+  uvwasi_destroy(&local_uvwasi_state);
+  wasm_rt_free();
```

Comment:
Wow, this uvwasi integration with wasm2c looks pretty nice... I guess we could extend this and split it out at some point since its seems generally useful for running standalone wasi programs under wasm2c?

Reply:
Yup, I've been meaning to fix the uvwasi integration for wasm2c for a while... Happy to add this feature to my todo list :)